src/ray/object_manager/ownership_based_object_directory.cc

```diff
@@ -95,101 +98,106 @@ rpc::Address GetOwnerAddressFromObjectInfo(const ObjectInfo &object_info) {
-  WorkerID worker_id = WorkerID::FromBinary(owner_address.worker_id());
-  if (worker_id.IsNil()) {
+  if (WorkerID::FromBinary(owner_address.worker_id()).IsNil()) {
     // If an object does not have owner, return nullptr.
     return nullptr;
   }
-  auto it = worker_rpc_clients_.find(worker_id);
-  if (it == worker_rpc_clients_.end()) {
-    it = worker_rpc_clients_
-             .emplace(worker_id, std::make_shared<rpc::CoreWorkerClient>(
-                                     owner_address, client_call_manager_))
-             .first;
-  }
-  return it->second;
+  return owner_client_pool_->GetOrConnect(owner_address);
 }
 
-ray::Status OwnershipBasedObjectDirectory::ReportObjectAdded(
-    const ObjectID &object_id, const NodeID &node_id, const ObjectInfo &object_info) {
+void OwnershipBasedObjectDirectory::ReportObjectAdded(const ObjectID &object_id,
+                                                      const NodeID &node_id,
+                                                      const ObjectInfo &object_info) {
   const WorkerID &worker_id = object_info.owner_worker_id;
-  rpc::Address owner_address = GetOwnerAddressFromObjectInfo(object_info);
-  std::shared_ptr<rpc::CoreWorkerClient> rpc_client = GetClient(owner_address);
-  if (rpc_client == nullptr) {
+  const auto owner_address = GetOwnerAddressFromObjectInfo(object_info);
+  auto owner_client = GetClient(owner_address);
+  if (owner_client == nullptr) {
     RAY_LOG(DEBUG) << "Object " << object_id << " does not have owner. "
                    << "ReportObjectAdded becomes a no-op."
                    << "This should only happen for Plasma store warmup objects.";
-    return Status::OK();
+    return;
   }
-  rpc::AddObjectLocationOwnerRequest request;
-  request.set_intended_worker_id(object_info.owner_worker_id.Binary());
-  request.set_object_id(object_id.Binary());
-  request.set_node_id(node_id.Binary());
-
   metrics_num_object_locations_added_++;
-
-  auto operation = [rpc_client, request, worker_id, object_id,
-                    node_id](const SequencerDoneCallback &done_callback) {
-    rpc_client->AddObjectLocationOwner(
-        request, [worker_id, object_id, node_id, done_callback](
-                     Status status, const rpc::AddObjectLocationOwnerReply &reply) {
-          if (!status.ok()) {
-            RAY_LOG(DEBUG) << "Worker " << worker_id << " failed to add the location "
-                           << node_id << " for " << object_id
-                           << ", the object has most likely been freed: "
-                           << status.ToString();
-          } else {
-            RAY_LOG(DEBUG) << "Added location " << node_id << " for object " << object_id
-                           << " on owner " << worker_id;
-          }
-          done_callback();
-        });
-  };
-  sequencer_.Post(object_id, operation);
-  return Status::OK();
+  location_buffers_[worker_id][object_id] = rpc::ObjectLocationState::ADDED;
```

Comment:
does the same owner_worker_id always map to the same owner address? (for example during the reconstruction of lost object)

Reply:
I believe so. Our reconstruction works only when the owner is alive, so I don't think the address is changed in that case? (Lmk if you mean sth else)

Afaik it is that; 

worker ID -> always have one IP
one IP -> can have multiple worker IDs 

If the owner address is not associated with the worker ID, that's handled in the core worker.cc